### PR TITLE
[4.0] Fix JS errors for com_joomlaupdate

### PR DESCRIPTION
**Pull request by joomdonation**
Pull Request for Issue # .

### Summary of Changes
There are currently many js errors in com_joomlaupdate. This PR attempts to fix this.

### Testing Instructions
I'm unsure as I don't understand fully how pre-update checks in com_joomlaupdate works yet. Maybe for now, code review would be enough because the current code is already broken.

### CODE REVIEW

File: build/media_source/com_joomlaupdate/js/default.es6.js [hunk 1]
```diff
@@ -495,10 +495,12 @@ Joomla = window.Joomla || {};
               >
               <span class="icon-help"></span>
               ${Joomla.JText._('COM_JOOMLAUPDATE_VIEW_DEFAULT_HELP')}
-              </span>`);
-          const popoverElement = pluginTitleTableCell.find('.hasPopover');
-          popoverElement.style.cursor = 'pointer';
-          popoverElement.popover({ placement: 'top', trigger: 'focus click' });
+              </span>`;
+          const popoverElement = pluginTitleTableCell.querySelector('.hasPopover');
+          if (popoverElement) {
```
Comment (joomdonation) [inline build/media_source/com_joomlaupdate/js/default.es6.js:495]: @dgrammatiko Do you know how to convert this block of code? The original code from 3.10.0 can be found here https://github.com/joomla/joomla-cms/blob/3.10-dev/media/com_joomlaupdate/js/default.js#L331-L333 . Not sure how to do this conversion on Joomla 4?

**Comment by richard67** [comment]
@joomdonation It needs to fix javascript-cs, see https://ci.joomla.org/joomla/joomla-cms/42929/1/21 .

File: build/media_source/com_joomlaupdate/js/default.es6.js (hunk 1 above)
Comment (richard67) [reply build/media_source/com_joomlaupdate/js/default.es6.js:495]: Maybe use `let` instead of `const` so you get:

```
let popoverElement = pluginTitleTableCell.querySelector('.hasPopover');
if (popoverElement) {
  popoverElement.style.cursor = 'pointer';
  popoverElement.popover({placement: 'top', trigger: 'focus click'});
}
```

Comment (joomdonation) [reply build/media_source/com_joomlaupdate/js/default.es6.js:495]: I think it is not related. I don't know how to add popover to the element here, so @dgrammatiko  or @Fedik or maybe @wilsonge could help us.

Comment (dgrammatiko) [reply build/media_source/com_joomlaupdate/js/default.es6.js:495]: To initialize the popover use

```js
new bootstrap.Popover(popoverElement, {placement: 'top', trigger: 'focus click'});
```

Comment (dgrammatiko) [reply build/media_source/com_joomlaupdate/js/default.es6.js:495]: Btw you need to include the popover in the layout HTMLHelper::_('bootstrap.popover')

File: build/media_source/com_joomlaupdate/js/default.es6.js [hunk 2]
```diff
@@ -454,11 +454,10 @@ Joomla = window.Joomla || {};
 
     // Process the nonCoreCriticalPlugin list
     if (extensionData.compatibilityData.resultGroup === 3) {
-      PreUpdateChecker.nonCoreCriticalPlugins.foreach((cpi) => {
-        // TODO: Make this typesafe
-        if (PreUpdateChecker.nonCoreCriticalPlugins[cpi].package_id === extensionId
-            || PreUpdateChecker.nonCoreCriticalPlugins[cpi].extension_id === extensionId) {
-          document.getElementById(`#plg_${PreUpdateChecker.nonCoreCriticalPlugins[cpi].extension_id}`).remove();
+      PreUpdateChecker.nonCoreCriticalPlugins.forEach((plugin, cpi) => {
+        if (plugin.package_id.toString() === extensionId
```
Comment (wilsonge) [inline build/media_source/com_joomlaupdate/js/default.es6.js:454]: Does toString work if it's acting on a string already (i'm sure there was a database type where integers' come out as a string)

File: build/media_source/com_joomlaupdate/js/default.es6.js (hunk 1 above)
Comment (wilsonge) [reply build/media_source/com_joomlaupdate/js/default.es6.js:495]: > Btw you need to include the popover in the layout HTMLHelper::_('bootstrap.popover')

I called WA addScript directly in the layout - because I don't think we need the HtmlHelper initialisation of the element?

File: build/media_source/com_joomlaupdate/js/default.es6.js (hunk 2 above)
Comment (Fedik) [reply build/media_source/com_joomlaupdate/js/default.es6.js:454]: yes it will work, `"blabla".toString()`

File: build/media_source/com_joomlaupdate/js/default.es6.js (hunk 1 above)
Comment (joomdonation) [reply build/media_source/com_joomlaupdate/js/default.es6.js:495]: @dgrammatiko Thanks, implemented and confirm that it works.

File: administrator/components/com_joomlaupdate/tmpl/joomlaupdate/default_preupdatecheck.php [hunk 3]
```diff
@@ -19,7 +19,7 @@
 // JText::script doesn't have a sprintf equivalent so work around this
 Factory::getDocument()->addScriptDeclaration("var COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSIONS_SHOW_MORE_COMPATIBILITY_INFORMATION = '" . Text::sprintf('COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSIONS_SHOW_MORE_COMPATIBILITY_INFORMATION', '<span class="icon-chevron-right large-icon" style="font-size:0.85rem"></span>', true) . "';");
```
Comment (dgrammatiko) [inline administrator/components/com_joomlaupdate/tmpl/joomlaupdate/default_preupdatecheck.php]: Can we use also `$doc->addScriptOptions()` for these 2 strings here?

I kinda hate the var and the inline js just to pass data from PHP to the JS world, so 2005...

File: administrator/components/com_joomlaupdate/tmpl/joomlaupdate/default_preupdatecheck.php [hunk 4]
```diff
@@ -19,7 +19,7 @@
...
 Factory::getDocument()->addScriptDeclaration("var COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSIONS_SHOW_LESS_COMPATIBILITY_INFORMATION = '" . Text::sprintf('COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSIONS_SHOW_LESS_COMPATIBILITY_INFORMATION', '<span class="icon-chevron-up large-icon" style="font-size:0.85rem"></span>', true) . "';");
```
Comment (dgrammatiko) [inline administrator/components/com_joomlaupdate/tmpl/joomlaupdate/default_preupdatecheck.php]: Same as above...

File: administrator/components/com_joomlaupdate/tmpl/joomlaupdate/default_preupdatecheck.php (hunk 3 above)
Comment (joomdonation) [reply administrator/components/com_joomlaupdate/tmpl/joomlaupdate/default_preupdatecheck.php]: Original code :).

Comment (dgrammatiko) [reply administrator/components/com_joomlaupdate/tmpl/joomlaupdate/default_preupdatecheck.php]: I know and probably should have made that comment the other day but would be great if we could "fix" it in this PR. If not no worries

**Comment by joomdonation** [comment]
@wilsonge As of right now, all the errors are fixed and I would say that it works in the same way with what we are having on 3.10.0. You can review the code and get it merged (I made basic test by installing older version of Akeeba Backup and JCE editor for testing)

However, there are still some problems:

- The layout of the page does not look good. We should have someone with frentend/design skill to take a look at it and sovle it
- During the code review, I can see a potential issue (comes from the original code, before we do this conversion to convert to vanilla js). See https://github.com/joomla/joomla-cms/blob/3.10-dev/media/com_joomlaupdate/js/default.js#L292

Basically, from what I understand, the loop over all the none core plugins and if that plugin belong to an extension, remove it from the array. But that could cause all plugins load being checked (causes by the removal mentioned). Maybe you can ask the developer who worked on the feature to take a look at it.

**Comment by dgrammatiko** [comment]
> However, there are still some problems:

George had asked me to fix this after the BS js PR but I never made a PR because my solution (SPA basically)was touching the restore.php and that wasn't a discussion I wanted myself involved. This is (I think) the repo: https://github.com/joomla/joomla-cms/compare/4.0-dev...dgrammatiko:4.0-dev_joomlaUpdate?expand=1

At the very least check the changes in the controller and the way I pass data from PHP to JS. The current state is god awful...

**Comment by joomdonation** [comment]
@dgrammatiko I haven't spent time to read the whole code to understand the process, so I could not comment on your solution right now. However, if you make a PR, I would be happy to review and test it. The current code is not looking good, I agree.

File: administrator/components/com_joomlaupdate/tmpl/joomlaupdate/default_preupdatecheck.php (hunk 3 above)
Comment (joomdonation) [reply administrator/components/com_joomlaupdate/tmpl/joomlaupdate/default_preupdatecheck.php]: @dgrammatiko Fixed that as suggested :)

File: build/media_source/com_joomlaupdate/js/default.es6.js [hunk 5]
```diff
@@ -474,31 +473,33 @@ Joomla = window.Joomla || {};
...
               ${Joomla.JText._('COM_JOOMLAUPDATE_VIEW_DEFAULT_HELP')}
-              </span>`);
-          const popoverElement = pluginTitleTableCell.find('.hasPopover');
-          popoverElement.style.cursor = 'pointer';
-          popoverElement.popover({ placement: 'top', trigger: 'focus click' });
+              </span>`;
+          const popoverElement = pluginTitleTableCell.querySelector('.hasPopover');
+          if (popoverElement) {
+            popoverElement.style.cursor = 'pointer';
+            const popover = new bootstrap.Popover(popoverElement, { placement: 'top', html: true, trigger: 'focus click' });
```
Comment (joomdonation) [inline build/media_source/com_joomlaupdate/js/default.es6.js]: @dgrammatiko Could you help with the CS issue related to this line of code ? Leave it like that, the CS error is https://ci.joomla.org/joomla/joomla-cms/43002/1/21 . If I change it to the code below, another CS error appear
```javascript
new bootstrap.Popover(popoverElement, { placement: 'top', html: true, trigger: 'focus click' });
```
Don't know how to make it happy :(

Comment (dgrammatiko) [reply build/media_source/com_joomlaupdate/js/default.es6.js]: What's the error when you're using this:

```js
new bootstrap.Popover(popoverElement, { placement: 'top', html: true, trigger: 'focus click' });
```

Comment (joomdonation) [reply build/media_source/com_joomlaupdate/js/default.es6.js]: I don't remember exactly but it does throws a message like you cannot create a new object without storing it in a variable :(.

Comment (dgrammatiko) [reply build/media_source/com_joomlaupdate/js/default.es6.js]: Just mute it using: `// eslint-disable-next-line no-new` 

Unless this `bootstrap.Popover(popoverElement, { placement: 'top', html: true, trigger: 'focus click' });` works (I doubt it will)

Comment (joomdonation) [reply build/media_source/com_joomlaupdate/js/default.es6.js]: Thanks @dgrammatiko . Had to mute it using `// eslint-disable-next-line no-new` as instructed.

File: build/media_source/com_joomlaupdate/js/default.es6.js [hunk 6]
```diff
@@ -140,14 +140,14 @@ Joomla = window.Joomla || {};
         if (toggle.target.dataset.state === 'closed') {
           toggle.target.dataset.state = 'open';
           // eslint-disable-next-line max-len,no-undef
```
Comment (wilsonge) [inline build/media_source/com_joomlaupdate/js/default.es6.js]: You can remove the `no-undef` ignore here (possible max-len too now this is a proper function - but less sure - same in the other places too where you're using the getOptions)

**Comment by wilsonge** [comment]
Starting to look better. Thanks for fixing my mess trying to port this!

File: build/media_source/com_joomlaupdate/js/default.es6.js (hunk 6 above)
Comment (joomdonation) [reply build/media_source/com_joomlaupdate/js/default.es6.js]: @wilsonge Done.

**Comment by wilsonge** [comment]
This is clearly an improvement merging this now
